drivers/dma/dma_mcux_smartdma.c

```diff
@@ -0,0 +1,251 @@
+		param.buffersize = config->head_block->block_size;
+		param.smartdma_stack = data->smartdma_stack;
+		/* Save configuration to SMARTDMA */
+		dev_config->base->ARM2SMARTDMA = (uint32_t)(&param);
+		break;
+	default:
+		/* For the default case, we simply pass the entire DMA config
+		 * struct to the SMARTDMA. The user's application could then
+		 * parse this structure, or choose to write different
+		 * configuration data to the SMARTDMA in their application
+		 */
+		dev_config->base->ARM2SMARTDMA = ((uint32_t)config);
+		break;
+	}
+	/* Save program */
+	dev_config->base->BOOT = (uint32_t)dev_config->smartdma_progs[prog_idx];
+	LOG_DBG("Boot address set to 0x%X", dev_config->base->BOOT);
+	return 0;
+}
+
+static int dma_mcux_smartdma_start(const struct device *dev, uint32_t channel)
+{
+	const struct dma_mcux_smartdma_config *config = dev->config;
+
+#ifdef CONFIG_PM
+	/* Block PM transition until DMA completes */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+#endif
+	/* Kick off SMARTDMA */
+	config->base->CTRL = SMARTDMA_MAGIC | SMARTDMA_BOOT;
+	return 0;
+}
+
+
+static int dma_mcux_smartdma_stop(const struct device *dev, uint32_t channel)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channel);
+	/* Stop DMA  */
+	SMARTDMA_Reset();
+#ifdef CONFIG_PM
+	/* Release PM lock */
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+#endif
+	return 0;
+}
+
+static int dma_mcux_smartdma_init(const struct device *dev)
+{
+	const struct dma_mcux_smartdma_config *config = dev->config;
+	/*
+	 * Initialize the SMARTDMA with firmware. The default firmware
+	 * from MCUX SDK is a display firmware, which has functions
+	 * implemented above in the dma configuration function. The
+	 * user can install another firmware using `dma_smartdma_install_fw`
+	 */
+	SMARTDMA_Init((uint32_t)config->smartdma_progs,
+		      s_smartdmaDisplayFirmware,
+		      SMARTDMA_DISPLAY_FIRMWARE_SIZE);
+	config->irq_config_func(dev);
+
+	return 0;
+}
+
+static void dma_mcux_smartdma_irq(const struct device *dev)
+{
+	const struct dma_mcux_smartdma_data *data = dev->data;
+
+	if (data->callback) {
+		data->callback(dev, data->user_data, 0, 0);
+	}
+#ifdef CONFIG_PM
+	/* Release PM lock */
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+#endif
+}
+
+/**
+ * @brief install SMARTDMA firmware
+ *
+ * Install a custom firmware for the smartDMA. This function allows the user
+ * to install a custom firmware into the smartDMA, which implements
+ * different API functions than the standard MCUX SDK firmware.
+ * @param dev: smartDMA device
+ * @param firmware: address of buffer containing smartDMA firmware
+ * @param len: length of firmware buffer
+ */
+void dma_smartdma_install_fw(const struct device *dev, uint8_t *firmware,
+			     uint32_t len)
+{
+	const struct dma_mcux_smartdma_config *config = dev->config;
+
+	SMARTDMA_InstallFirmware((uint32_t)config->smartdma_progs, firmware, len);
```

Comment:
what's the point of having this device custom wrapper function?

Reply:
The wrapper function uses the `smartdma_progs` pointer as the location to install the firmware to. This way if a custom firmware was used, it could be loaded using this wrapper function, then a user would call the `dma_config` API with a slot number based on what the custom firmware required

Effectively, the advantage of this wrapper function is that it loads the SMARTDMA firmware to the same offset as is specified in devicetree and used by the `dma_config` function when instructing the SMARTDMA on what operation to perform